### PR TITLE
Enable xunit analyzers

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Enable xunit.analyzers for our repo, and fix up some of the more egregious issues. The ones that are "less bad" are disabled, and will be dealt with in follow-up PRs to keep this PR manageable.

### CODE REVIEW

File: build/Rulesets/Roslyn_BuildRules.ruleset [hunk 1]
```diff
@@ -123,4 +123,15 @@
   <Rules AnalyzerId="XmlDocumentationComments.Analyzers" RuleNamespace="XmlDocumentationComments.Analyzers">
     <Rule Id="RS0010" Action="Warning" />
   </Rules>
+  <Rules AnalyzerId="xunit.analyzers" RuleNamespace="Xunit.Analyzers">
+    <Rule Id="xUnit1004" Action="None" /> <!-- allow skipped tests -->
+    <Rule Id="xUnit1024" Action="None" /> <!-- seems broken: https://github.com/xunit/xunit/issues/1361 -->
+	<Rule Id="xUnit2000" Action="None" /> <!-- "literal or constant value should be first in Assert.Equal" is a valid assert, but very noisy. Will do the refactoring later -->
```
Comment (jaredpar) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: Indentation difference here.

**Comment by jasonmalinowski** [comment]
(I can only assume that some of the tests I'm enabling are going to fail. Hopefully not due to product bugs!)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenThrowTests.cs [hunk 2]
```diff
@@ -81,6 +81,7 @@ .maxstack  1
 ");
         }
 
+        [Fact]
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenThrowTests.cs]: Yeah ... we must merge this PR. Too much of this.

File: src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs [hunk 3]
```diff
@@ -172,71 +172,71 @@ private void VerifyBreakIntoCharacterParts(string original, params string[] part
...
-        [InlineData("[|Co|]deFix_[|Pro|]vider", "copro", PatternMatchKind.CamelCaseNonContiguousPrefix, CaseInsensitive)]
-        [InlineData("[|C|]ore[|Ofi|]lac[|Pro|]fessional", "cofipro", PatternMatchKind.CamelCaseExact, CaseInsensitive)]
-        [InlineData("[|C|]lear[|Ofi|]lac[|Pro|]fessional", "cofipro", PatternMatchKind.CamelCaseExact, CaseInsensitive)]
-        [InlineData("[|CO|]DE_FIX_[|PRO|]VIDER", "copro", PatternMatchKind.CamelCaseNonContiguousPrefix, CaseInsensitive)]
-
-        [InlineData("my[|_b|]utton", "_B", PatternMatchKind.CamelCaseSubstring, CaseInsensitive)]
-        [InlineData("[|_|]my_[|b|]utton", "_B", PatternMatchKind.CamelCaseNonContiguousPrefix, CaseInsensitive)]
+        [InlineData("[|Foo|]", "Foo", (int)PatternMatchKind.Exact, CaseSensitive)]
+        [InlineData("[|foo|]", "Foo", (int)PatternMatchKind.Exact, CaseInsensitive)]
+        [InlineData("[|Foo|]", "foo", (int)PatternMatchKind.Exact, CaseInsensitive)]
```
Comment (jaredpar) [inline src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs]: What is the logic behind this change?

Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs]: There is an analyzer that enforces that the types in the InlineData match the types of the parameters. It looks like we were relying on that the enum was being converted to an int somewhere.

Comment (jaredpar) [inline src/EditorFeatures/Test/Utilities/PatternMatcherTests.cs]: Gotcha

File: build/Targets/Packages.props [hunk 4]
```diff
@@ -223,6 +223,7 @@
     <VsWebsiteInteropVersion>8.0.0.0-alpha</VsWebsiteInteropVersion>
     <vswhereVersion>1.0.71</vswhereVersion>
     <xunitVersion>2.2.0-beta4-build3444</xunitVersion>
+    <xunitanalyzersVersion>0.5.0</xunitanalyzersVersion>
```
Comment (agocke) [inline build/Targets/Packages.props]: Did you give any thought to just upgrading the `xunitVersion` to `2.2.0`? It now references the analyzers as part of the package.

Comment (jaredpar) [inline build/Targets/Packages.props]: @agocke i advised against that due to you saying newer xunit broke our CoreClr testing. Did I misunderstand you?

Comment (agocke) [inline build/Targets/Packages.props]: If we just upgrade the library, I think it should work.

Comment (jasonmalinowski) [inline build/Targets/Packages.props]: @agocke Do we actually have something written down anywhere describing the problem? Right now the answer seems to be "it's broken, but nobody knows why".

Comment (agocke) [inline build/Targets/Packages.props]: The problem I was having was with the new runners, specifically the "dotnet xunit" runner. The library itself I hit no issues with, as far as I can tell (aside from the new warnings produced by the analyzer, of course).

Comment (jasonmalinowski) [inline build/Targets/Packages.props]: @agocke Do we have a bug tracking somewhere?

Comment (agocke) [inline build/Targets/Packages.props]: @jasonmalinowski No, this was just something I was playing around with on my Mac.

**Comment by jasonmalinowski** [comment]
@jaredpar @mavasani I think I've cleaned this up so all current flagged items are fixed. Given the other churn around enabling/disabling analyzers right now, any reason I should hold off merging this?

**Comment by jaredpar** [comment]
I say merge. Most of the analyzer problems are related to our self hosting of IOperation. Shouldn't have issues with other analyzers.

**Pull request merged**